dask/tests/test_distributed.py

```diff
@@ -398,3 +335,93 @@ async def test_annotation_pack_unpack(c, s, a, b):
     unpacked_hlg = HighLevelGraph.__dask_distributed_unpack__(packed_hlg, annotations)
     annotations = unpacked_hlg["annotations"]
     assert annotations == {"workers": {"n": ("alice",)}}
+
+
+class SchedulerImportCheck(SchedulerPlugin):
+    """Plugin to help record which modules are imported on the scheduler"""
+
+    def __init__(self, pattern):
+        self.pattern = pattern
+
+    async def start(self, scheduler):
+        # Record the modules that have been imported when the scheduler starts
+        self.start_modules = set()
+        for mod in set(sys.modules):
+            if not mod.startswith(self.pattern):
+                self.start_modules.add(mod)
+            else:
+                # Maually remove the target library
+                sys.modules.pop(mod)
+
+
+def get_start_modules(dask_scheduler):
+    import_check_plugins = [
+        p for p in dask_scheduler.plugins if type(p) is SchedulerImportCheck
+    ]
+    assert len(import_check_plugins) == 1
+
+    plugin = import_check_plugins[0]
+    return plugin.start_modules
+
+
+def _dataframe_shuffle():
+    pd = pytest.importorskip("pandas")
+    dd = pytest.importorskip("dask.dataframe")
+
+    # Perform a computation using an HLG-based shuffle
+    df = pd.DataFrame({"a": range(10), "b": range(10, 20)})
+    return dd.from_pandas(df, npartitions=2)
+
+
+def _dataframe_broadcast_join():
+    pd = pytest.importorskip("pandas")
+    dd = pytest.importorskip("dask.dataframe")
+
+    # Perform a computation using an HLG-based broadcast join
+    df = pd.DataFrame({"a": range(10), "b": range(10, 20)})
+    ddf1 = dd.from_pandas(df, npartitions=4)
+    ddf2 = dd.from_pandas(df, npartitions=1)
+    return ddf1.merge(ddf2, how="left", broadcast=True)
+
+
+def _array_creation():
+    da = pytest.importorskip("dask.array")
+
+    # Perform a computation using HLG-based array creation
+    return da.ones((100,)) + da.zeros((100,))
+
+
+@pytest.mark.parametrize(
+    "op,lib",
+    [
+        (_dataframe_shuffle, "pandas."),
+        (_dataframe_broadcast_join, "pandas."),
+        (_array_creation, "numpy."),
+    ],
+)
+@pytest.mark.parametrize("optimize_graph", [True, False])
+def test_scheduler_highlevel_graph_unpack_import(op, lib, optimize_graph, loop):
+    # Test to highlight that in some cases we import modules like pandas
+    # on the scheduler when unpacking some HighLevelGraphs.
+    # This is a problem because pandas may not be installed in the process
+    # where the scheduler is running -- leading to an ImportError.
```

Comment:
@jrbourbeau - I tried working in your testing idea from #7374 here.  Let me know if you would like me to focus on testing in this PR, or if you would like the changes to go in that PR with `xfail` (and then we can simply remove the `xfail` in this PR)

Reply:
Including this here sounds great -- thanks